### PR TITLE
[`ChatGlm`] Adds support for the ChatGLM model

### PR DESCRIPTION
**Pull request by ArthurZucker**
# What does this PR do?
Drafts the support of chat GLM

### CODE REVIEW

**Comment by changwangss** [comment]
Hi~ o(*￣▽￣*)ブ Could  you tell me what your plans? May I know when the PR can be merged?  I have PRs depended on this feature.

**Comment by ArthurZucker** [comment]
@younesbelkada and I just came back from holidays, we are hoping end of the week maybe later!

File: src/transformers/models/chatglm/modeling_chatglm.py [hunk 1]
```diff
@@ -0,0 +1,1284 @@
...
+        position_ids: Optional[torch.LongTensor] = None,
+        past_key_value: Optional[Cache] = None,
+        output_attentions: bool = False,
+        use_cache: bool = False,
+    ) -> Tuple[torch.Tensor, Optional[torch.Tensor], Optional[Tuple[torch.Tensor]]]:
+        if output_attentions:
+            # TODO: Improve this warning with e.g. `model.config.attn_implementation = "manual"` once this is implemented.
+            logger.warning_once(
+                "ChatGlmModel is using ChatGlmSdpaAttention, but `torch.nn.functional.scaled_dot_product_attention` does not support `output_attentions=True`. Falling back to the manual attention implementation, "
+                'but specifying the manual implementation will be required from Transformers version v5.0.0 onwards. This warning can be removed using the argument `attn_implementation="eager"` when loading the model.'
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/modeling_chatglm.py:541]: should be split in two lines

File: src/transformers/models/chatglm/modeling_chatglm.py [hunk 2]
```diff
@@ -0,0 +1,1284 @@
...
+
+    def set_decoder(self, decoder):
+        self.model = decoder
+
+    def get_decoder(self):
+        return self.model
+
+    @add_start_docstrings_to_model_forward(CHATGLM_INPUTS_DOCSTRING)
+    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
+    # Ignore Copy
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/modeling_chatglm.py:1009]:
```suggestion
    # Ignore Copy
    @add_start_docstrings_to_model_forward(CHATGLM_INPUTS_DOCSTRING)
    @replace_return_docstrings(output_type=CausalLMOutputWithPast, config_class=_CONFIG_FOR_DOC)
```

Comment (ArthurZucker) [reply src/transformers/models/chatglm/modeling_chatglm.py]: Ignore copy

**Comment by younesbelkada** [comment]
cc @ArthurZucker could you give a first pass ? 🙏

File: docs/source/en/model_doc/chatglm.md [hunk 3]
```diff
@@ -0,0 +1,55 @@
+<!--Copyright 2023 The HuggingFace Team. All rights reserved.
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/chatglm.md:1]: Headers need an update

File: docs/source/en/model_doc/chatglm.md [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+The abstract from the paper is the following:
+
+*There have been various types of pretraining architectures including autoencoding models (e.g., BERT), autoregressive models (e.g., GPT), and encoder-decoder models (e.g., T5). However, none of the pretraining frameworks performs the best for all tasks of three main categories including natural language understanding (NLU), unconditional generation, and conditional generation. We propose a General Language Model (GLM) based on autoregressive blank infilling to address this challenge. GLM improves blank filling pretraining by adding 2D positional encodings and allowing an arbitrary order to predict spans, which results in performance gains over BERT and T5 on NLU tasks. Meanwhile, GLM can be pretrained for different types of tasks by varying the number and lengths of blanks. On a wide range of tasks across NLU, conditional and unconditional generation, GLM outperforms BERT, T5, and GPT given the same model sizes and data, and achieves the best performance from a single pretrained model with 1.25× parameters of BERTLarge, demonstrating its generalizability to different downstream tasks*
+
+Tips:
+
+- TODO conversion tips if needed
```
Comment (ArthurZucker) [inline docs/source/en/model_doc/chatglm.md:33]: to remove?

File: src/transformers/models/chatglm/__init__.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
+# Copyright 2023 EleutherAI and The HuggingFace Inc. team. All rights reserved.
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/__init__.py:1]: same here

File: src/transformers/models/chatglm/modeling_chatglm.py [hunk 6]
```diff
@@ -0,0 +1,1289 @@
...
+    cos = cos[position_ids].unsqueeze(unsqueeze_dim)
+    sin = sin[position_ids].unsqueeze(unsqueeze_dim)
+
+    batch_size, num_q_heads, seq_len, head_dim = q.shape
+    _, num_k_heads, _, _ = k.shape
+
+    rot_shape = cos.shape
+
+    # In the original ChatGLM repository the query and key states are manually
+    # reshaped into a shape `batch_size, num_q_heads, seq_len, head_dim // 2, 2` changing the order
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/modeling_chatglm.py:213]: don't you think that we can do this when converting the checkpoints?

File: src/transformers/models/chatglm/modeling_chatglm.py [hunk 7]
```diff
@@ -0,0 +1,1289 @@
...
+        self.partial_rotary_factor = config.partial_rotary_factor
+
+        self.rope_theta = config.rope_theta
+
+        if not self.multi_query_attention:
+            qkv_size = 3 * self.hidden_size
+            self.num_key_value_heads = 1
+        else:
+            self.num_key_value_heads = config.multi_query_group_num
+            qkv_size = self.projection_size + 2 * self.hidden_size_per_attention_head * self.num_key_value_heads
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/modeling_chatglm.py:360]: we should just have a general implementation! supporting GQA means you can have MQA, MHA and GQA

File: src/transformers/models/chatglm/modeling_chatglm.py [hunk 8]
```diff
@@ -0,0 +1,1289 @@
...
+                self.rotary_emb = ChatGlmDynamicNTKScalingRotaryEmbedding(
+                    int(self.partial_rotary_factor * self.head_dim),
+                    max_position_embeddings=self.max_position_embeddings,
+                    scaling_factor=scaling_factor,
+                    base=self.rope_theta,
+                )
+            else:
+                raise ValueError(f"Unknown RoPE scaling type {scaling_type}")
+
+    def _split_heads(self, fused_qkv: torch.Tensor) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor]:
```
Comment (ArthurZucker) [inline src/transformers/models/chatglm/modeling_chatglm.py:396]: we don't need if else

**Review by ArthurZucker — COMMENTED** [review]
A few nits here and there, ready otherwise! cc @younesbelkada

File: src/transformers/models/chatglm/modeling_chatglm.py (hunk 7 above)
Comment (ArthurZucker) [reply src/transformers/models/chatglm/modeling_chatglm.py:360]: agreed!

Comment (ArthurZucker) [reply src/transformers/models/chatglm/modeling_chatglm.py:360]: I might have time to work on the addition next week, with a new cool feature that will help ease the integration! 🤗

**Comment by ArthurZucker** [comment]
I think #33823 should fix this!

**Pull request closed without merging**